File: collector/internal/discover/discover.go
package discover

import (
	"os"
	"os/exec"
	"path/filepath"
	"raptor-collector/internal/axfr"
	"raptor-collector/internal/config"
	"raptor-collector/internal/namedconf"
	"raptor-collector/internal/zone"
	"runtime"
	"strings"
)

type Zone struct {
	Name      string
	Source    string
	FilePath  string
	SOASerial string
	Records   []zone.Record
}

var bindCandidates = []string{
	"/etc/bind/named.conf",
	"/etc/named.conf",
	"/etc/named/named.conf",
	"/chroot/etc/named.conf",
}

var pdnsCandidates = []string{
	"/etc/powerdns/pdns.conf",
	"/etc/pdns/pdns.conf",
}

func loadFile(path, origin string) []zone.Record {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	return zone.Parse(string(raw), origin)
}

func discoverBind(conf string) []Zone {
	candidates := bindCandidates
	if strings.TrimSpace(conf) != "" {
		candidates = []string{conf}
	}
	var found []Zone
	for _, path := range candidates {
		if _, err := os.Stat(path); err != nil {
			continue
		}
		for _, z := range namedconf.Load(path) {
			recs := []zone.Record{}
			if z.FilePath != "" {
				recs = loadFile(z.FilePath, z.Name)
			}
			found = append(found, Zone{Name: z.Name, Source: "bind", FilePath: z.FilePath, SOASerial: zone.SOASerial(recs), Records: recs})
		}
		if len(found) > 0 {
			break
		}
	}
	return found
}

func discoverPowerDNS(conf string) []Zone {
	path := conf
	if path == "" {
		for _, c := range pdnsCandidates {
			if _, err := os.Stat(c); err == nil {
				path = c
				break
			}
		}
	}
	if path == "" {
		return nil
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	bindConfig := ""
	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(strings.Split(line, "#")[0])
		if !strings.Contains(line, "=") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		key := strings.TrimSpace(parts[0])
		if key == "bind-config" || key == "bind-config-file" {
			bindConfig = strings.TrimSpace(parts[1])
		}
	}
	if bindConfig != "" {
		if _, err := os.Stat(bindConfig); err == nil {
			var found []Zone
			for _, z := range namedconf.Load(bindConfig) {
				recs := []zone.Record{}
				if z.FilePath != "" {
					recs = loadFile(z.FilePath, z.Name)
				}
				found = append(found, Zone{Name: z.Name, Source: "powerdns", FilePath: z.FilePath, SOASerial: zone.SOASerial(recs), Records: recs})
			}
			if len(found) > 0 {
				return found
			}
		}
	}
	out, err := exec.Command("pdnsutil", "list-all-zones").Output()
	if err != nil {
		return nil
	}
	var found []Zone
	for _, line := range strings.Split(string(out), "\n") {
		name := strings.ToLower(strings.TrimSuffix(strings.TrimSpace(line), "."))
		if name != "" {
			found = append(found, Zone{Name: name, Source: "powerdns"})
		}
	}
	return found
}

func discoverWindows() []Zone {
	if runtime.GOOS != "windows" {
		return nil
	}
	out, err := exec.Command("dnscmd", ".", "/EnumZones").Output()
	if err != nil {
		return nil
	}
	var found []Zone
	for _, line := range strings.Split(string(out), "\n") {
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		name := strings.ToLower(strings.TrimSuffix(fields[0], "."))
		if name == "" || name == "zone" || strings.Contains(strings.ToLower(name), "enumzone") {
			continue
		}
		if !strings.Contains(name, ".") && (name == "." || name == "command") {
			continue
		}
		z := Zone{Name: name, Source: "windows_dns"}
		printed, err := exec.Command("dnscmd", ".", "/ZonePrint", name).Output()
		if err == nil {
			recs := zone.Parse(string(printed), name)
			z.Records = recs
			z.SOASerial = zone.SOASerial(recs)
		}
		found = append(found, z)
	}
	return found
}

func discoverDirs(dirs []string) []Zone {
	var found []Zone
	for _, dir := range dirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, entry := range entries {
			if entry.IsDir() {
				continue
			}
			path := filepath.Join(dir, entry.Name())
			origin := strings.TrimSuffix(entry.Name(), ".db")
			origin = strings.ToLower(strings.TrimSuffix(origin, "."))
			recs := loadFile(path, origin)
			found = append(found, Zone{Name: origin, Source: "zone_dir", FilePath: path, SOASerial: zone.SOASerial(recs), Records: recs})
		}
	}
	return found
}

func All(cfg *config.Config) []Zone {
	var found []Zone
	found = append(found, discoverBind(cfg.BindConf)...)
	found = append(found, discoverPowerDNS(cfg.PowerDNSConf)...)
	found = append(found, discoverWindows()...)
	found = append(found, discoverDirs(cfg.ZoneDirs)...)
	byName := map[string]Zone{}
	for _, z := range found {
		if z.Name == "" {
			continue
		}
		if _, ok := byName[z.Name]; !ok {
			byName[z.Name] = z
		}
	}
	out := make([]Zone, 0, len(byName))
	for _, z := range byName {
		if len(z.Records) == 0 && cfg.AXFRServer != "" {
			if recs, err := axfr.Transfer(z.Name, cfg.AXFRServer, cfg.TSIGName, cfg.TSIGSecret, cfg.TSIGAlgorithm); err == nil {
				z.Records = recs
				z.SOASerial = zone.SOASerial(recs)
			}
		}
		out = append(out, z)
	}
	return out
}

func Payloads(zones []Zone) []map[string]any {
	out := make([]map[string]any, 0, len(zones))
	for _, z := range zones {
		out = append(out, map[string]any{
			"name":       z.Name,
			"soa_serial": z.SOASerial,
			"records":    zone.ToPayload(z.Records),
		})
	}
	return out
}

File: collector/raptor_collector/discover.py
"""Discover BIND, PowerDNS, and Windows DNS zone sources on this host."""

import logging
import os
import shutil
import subprocess
from dataclasses import dataclass, field
from typing import Iterable, List, Optional

from raptor_collector.named_conf import load_bind_zones
from raptor_collector.zone import ResourceRecord, parse_zone_records, soa_serial

logger = logging.getLogger(__name__)

BIND_CONF_CANDIDATES = (
    "/etc/bind/named.conf",
    "/etc/named.conf",
    "/etc/named/named.conf",
    "/chroot/etc/named.conf",
)
POWERDNS_CONF_CANDIDATES = (
    "/etc/powerdns/pdns.conf",
    "/etc/pdns/pdns.conf",
)


@dataclass
class DiscoveredZone:
    name: str
    source: str
    file_path: Optional[str] = None
    soa_serial: Optional[str] = None
    records: List[ResourceRecord] = field(default_factory=list)


def _read_text(path: str) -> str:
    with open(path, "r", encoding="utf-8", errors="replace") as handle:
        return handle.read()


def _load_zone_file(path: str, origin: str) -> List[ResourceRecord]:
    try:
        return parse_zone_records(_read_text(path), origin)
    except OSError as exc:
        logger.warning("Failed to read zone file %s: %s", path, exc)
        return []


def discover_bind(conf_path: str = "") -> List[DiscoveredZone]:
    candidates = (conf_path,) if conf_path else BIND_CONF_CANDIDATES
    found: List[DiscoveredZone] = []
    for path in candidates:
        if not path or not os.path.isfile(path):
            continue
        for zone in load_bind_zones(path):
            records = _load_zone_file(zone.file_path, zone.name) if zone.file_path else []
            found.append(
                DiscoveredZone(
                    name=zone.name,
                    source="bind",
                    file_path=zone.file_path,
                    soa_serial=soa_serial(records),
                    records=records,
                )
            )
        if found:
            break
    return found


def _parse_pdns_conf(path: str) -> dict:
    config = {}
    try:
        for line in _read_text(path).splitlines():
            stripped = line.split("#", 1)[0].strip()
            if not stripped or "=" not in stripped:
                continue
            key, value = stripped.split("=", 1)
            config[key.strip()] = value.strip()
    except OSError:
        return {}
    return config


def discover_powerdns(conf_path: str = "") -> List[DiscoveredZone]:
    path = conf_path or next((item for item in POWERDNS_CONF_CANDIDATES if os.path.isfile(item)), "")
    if not path:
        return []
    config = _parse_pdns_conf(path)
    bind_config = config.get("bind-config") or config.get("bind-config-file")
    zones: List[DiscoveredZone] = []
    if bind_config and os.path.isfile(bind_config):
        for zone in load_bind_zones(bind_config):
            records = _load_zone_file(zone.file_path, zone.name) if zone.file_path else []
            zones.append(
                DiscoveredZone(
                    name=zone.name,
                    source="powerdns",
                    file_path=zone.file_path,
                    soa_serial=soa_serial(records),
                    records=records,
                )
            )
        if zones:
            return zones

    pdnsutil = shutil.which("pdnsutil")
    if not pdnsutil:
        return zones
    try:
        listed = subprocess.run(
            [pdnsutil, "list-all-zones"],
            check=False,
            capture_output=True,
            text=True,
            timeout=15,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        logger.warning("pdnsutil list-all-zones failed: %s", exc)
        return zones
    if listed.returncode != 0:
        return zones
    for line in listed.stdout.splitlines():
        name = line.strip().rstrip(".").lower()
        if not name:
            continue
        zones.append(DiscoveredZone(name=name, source="powerdns"))
    return zones


def discover_windows_dns() -> List[DiscoveredZone]:
    if os.name != "nt":
        return []
    dnscmd = shutil.which("dnscmd")
    if not dnscmd:
        return []
    try:
        listed = subprocess.run(
            [dnscmd, ".", "/EnumZones"],
            check=False,
            capture_output=True,
            text=True,
            timeout=20,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        logger.warning("dnscmd /EnumZones failed: %s", exc)
        return []
    zones: List[DiscoveredZone] = []
    for line in listed.stdout.splitlines():
        parts = line.split()
        if len(parts) < 2:
            continue
        if parts[0].lower() in {"zone name", "zone"}:
            continue
        name = parts[0].strip().rstrip(".").lower()
        if "." not in name and name in {".", "enumzones", "command"}:
            continue
        if name:
            zones.append(DiscoveredZone(name=name, source="windows_dns"))
    return zones


def discover_zone_dirs(directories: Iterable[str]) -> List[DiscoveredZone]:
    found: List[DiscoveredZone] = []
    for directory in directories:
        if not directory or not os.path.isdir(directory):
            continue
        for name in os.listdir(directory):
            path = os.path.join(directory, name)
            if not os.path.isfile(path):
                continue
            origin = name
            if origin.endswith(".db"):
                origin = origin[:-3]
            origin = origin.rstrip(".").lower()
            records = _load_zone_file(path, origin)
            found.append(
                DiscoveredZone(
                    name=origin,
                    source="zone_dir",
                    file_path=path,
                    soa_serial=soa_serial(records),
                    records=records,
                )
            )
    return found


def discover_all(
    *,
    bind_conf: str = "",
    powerdns_conf: str = "",
    extra_zone_dirs: Optional[Iterable[str]] = None,
) -> List[DiscoveredZone]:
    discovered: List[DiscoveredZone] = []
    discovered.extend(discover_bind(bind_conf))
    discovered.extend(discover_powerdns(powerdns_conf))
    discovered.extend(discover_windows_dns())
    discovered.extend(discover_zone_dirs(extra_zone_dirs or ()))
    by_name = {}
    for zone in discovered:
        if zone.name and zone.name not in by_name:
            by_name[zone.name] = zone
    return list(by_name.values())
